### ISSUE TITLE
[KSP] How to pass compiler argument to KSP?

### ISSUE THREAD

**Issue by mpaduchhh** [google/dagger#4349]
Hey,
I am trying to pass fastInit flag to dagger, and I have tried every option I could find online, but I dont even know if its working.
Here are my questions:
- What is the recommended way of passing compiler argument when using ksp and gradle for multi module project?
- How to check if its working? I searched build logs for "dagger" keyword, but theres nothing there.

What I have tried:

```
allProjects {
    tasks.withType<JavaCompile> {
        options.compilerArgs.add("-Adagger.fastInit=enabled")

    }
    tasks.withType<KspTask> {
        options.add(SubpluginOption("-Adagger.fastInit", "enabled"))
        commandLineArgumentProviders.add(CommandLineArgumentProvider{
            listOf("-Adagger.fastInit=enabled",)
        })
    }
}
```

as well as putting this argument in `gradle.properties`

**Comment by bcorso** [comment]
See https://kotlinlang.org/docs/ksp-quickstart.html#pass-options-to-processors

**Comment by mpaduchhh** [comment]
Sorry, will need clarification on argument format. It's not clear from the guide.
I tried passing it without `-` and `A` and I am getting dagger errors, on unrelated flag:
I found this format in camera2 google repo.
```
e: [ksp] java.lang.NullPointerException: Cannot invoke "dagger.internal.codegen.validation.ValidationBindingGraphPlugins.initializePlugins()" because "this.validationBindingGraphPlugins" is null
        at dagger.internal.codegen.DelegateComponentProcessor.processingSteps(DelegateComponentProcessor.java:112)
        at dagger.internal.codegen.KspComponentProcessor.processingSteps(KspComponentProcessor.java:58)
        at dagger.spi.internal.shaded.androidx.room.compiler.processing.ksp.KspBasicAnnotationProcessor$steps$2.invoke(KspBasicAnnotationProcessor.kt:46)
        at dagger.spi.internal.shaded.androidx.room.compiler.processing.ksp.KspBasicAnnotationProcessor$steps$2.invoke(KspBasicAnnotationProcessor.kt:46)
        at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
        at dagger.spi.internal.shaded.androidx.room.compiler.processing.ksp.KspBasicAnnotationProcessor.getSteps(KspBasicAnnotationProcessor.kt:46)
        at dagger.spi.internal.shaded.androidx.room.compiler.processing.ksp.KspBasicAnnotationProcessor.access$getSteps(KspBasicAnnotationProcessor.kt:34)
        at dagger.spi.internal.shaded.androidx.room.compiler.processing.ksp.KspBasicAnnotationProcessor$commonDelegate$2.invoke(KspBasicAnnotationProcessor.kt:48)
        at dagger.spi.internal.shaded.androidx.room.compiler.processing.ksp.KspBasicAnnotationProcessor$commonDelegate$2.invoke(KspBasicAnnotationProcessor.kt:48)
        at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
        at dagger.spi.internal.shaded.androidx.room.compiler.processing.ksp.KspBasicAnnotationProcessor.getCommonDelegate(KspBasicAnnotationProcessor.kt:48)
        at dagger.spi.internal.shaded.androidx.room.compiler.processing.ksp.KspBasicAnnotationProcessor.runLastRound(KspBasicAnnotationProcessor.kt:81)
        at dagger.spi.internal.shaded.androidx.room.compiler.processing.ksp.KspBasicAnnotationProcessor.onError(KspBasicAnnotationProcessor.kt:76)
        at com.google.devtools.ksp.AbstractKotlinSymbolProcessingExtension$doAnalysis$11$1.invoke(KotlinSymbolProcessingExtension.kt:332)
        at com.google.devtools.ksp.AbstractKotlinSymbolProcessingExtension$doAnalysis$11$1.invoke(KotlinSymbolProcessingExtension.kt:331)
        at com.google.devtools.ksp.AbstractKotlinSymbolProcessingExtension.handleException(KotlinSymbolProcessingExtension.kt:410)
        at com.google.devtools.ksp.AbstractKotlinSymbolProcessingExtension.doAnalysis(KotlinSymbolProcessingExtension.kt:331)
        at org.jetbrains.kotlin.cli.jvm.compiler.TopDownAnalyzerFacadeForJVM.analyzeFilesWithJavaIntegration(TopDownAnalyzerFacadeForJVM.kt:112)
        at org.jetbrains.kotlin.cli.jvm.compiler.TopDownAnalyzerFacadeForJVM.analyzeFilesWithJavaIntegration$default(TopDownAnalyzerFacadeForJVM.kt:77)
        at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler$analyze$1.invoke(KotlinToJVMBytecodeCompiler.kt:256)
        at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler$analyze$1.invoke(KotlinToJVMBytecodeCompiler.kt:247)
        at org.jetbrains.kotlin.cli.common.messages.AnalyzerWithCompilerReport.analyzeAndReport(AnalyzerWithCompilerReport.kt:115)
        at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.analyze(KotlinToJVMBytecodeCompiler.kt:247)
        at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.compileModules$cli(KotlinToJVMBytecodeCompiler.kt:87)
        at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.compileModules$cli$default(KotlinToJVMBytecodeCompiler.kt:43)
        at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:165)
        at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:50)
        at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:104)
        at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:48)
        at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:101)
        at org.jetbrains.kotlin.daemon.CompileServiceImpl.compile(CompileServiceImpl.kt:1523)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at java.rmi/sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:360)
        at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:200)
        at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:197)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:712)
        at java.rmi/sun.rmi.transport.Transport.serviceCall(Transport.java:196)
        at java.rmi/sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:587)
        at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:828)
        at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:705)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
        at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:704)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:840)

```


When passing with `- `and/or `A` build passes, but I dont see effect of my flags.



**Could u advise how to confirm if the flag is applied correctly?**

**Comment by bcorso** [comment]
> Sorry, will need clarification on argument format. It's not clear from the guide.

```
ksp {
    arg("dagger.fastInit", "enabled")
}
```

> I tried passing it without - and A and I am getting dagger errors, on unrelated flag

I don't really know why that would be failing (it should never be null). If you still get this with the setup above I would try using K1 rather than K2 because Dagger doesn't officially support K2 yet so there may be some unexpected issues (I'm assuming you're using K2 from the `K2JVMCompiler` in the error message).

> Could u advise how to confirm if the flag is applied correctly?

The `fastInit` flag changes the generated code for the component, so you would need to look at the generated code. Generally, you can tell if you are using `fastInit` based on if you see `SwitchingProvider` in the generated code. Feel free to post your generated code if you have doubts.

**Comment by mpaduchhh** [comment]
Thanks. I solved the issue. Since, I didn't know what change to expect from generated code I changed the flag from fastInit to format to see some obvious changes and actually, the error was coming from the format flag, for whatever reason.
The fastInit works fine with the format you posted. 
If anyone is looking for a way to enable it for every module in multi module gradle project with ksp, below is the way:

```
allProjects {
   tasks.withType<KspTask> {
        commandLineArgumentProviders.add(CommandLineArgumentProvider{
            listOf("dagger.fastInit=enabled")
        })
    }
}
```
